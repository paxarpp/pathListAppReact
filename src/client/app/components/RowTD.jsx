import React from 'react';
import PropTypes from 'prop-types';

import createObjectError from './createObjectError';

const RowTD = ({ path, handler, error, doubleClick }) => {
  const matchNames = createObjectError(path, error);
  return (
    <tr onClick={path.name !== null ? handler(path) : null}>
      <td>{path.name}</td>
      {path.dateBegin === null ? <td></td> : <td>{new Date(path.dateBegin).toLocaleDateString()}</td>}
      <td>{path.fuel}</td>
      <td>{path.constFuelChange}</td>
      <td
        className={!matchNames.first && matchNames.path ? 'inputError' : null}
        onDoubleClick={
          path.name !== null ? doubleClick('pathBegin', path.pathBegin) : null
        }
      >
        {path.name !== null ? (
          <span className="editableTd">{path.pathBegin}</span>
        ) : null}
      </td>
      <td
        className={matchNames.first && matchNames.path ? 'inputError' : null}
        onDoubleClick={
          path.name !== null ? doubleClick('pathEnd', path.pathEnd) : null
        }
      >
        {path.name !== null ? (
          <span className="editableTd">{path.pathEnd}</span>
        ) : null}
      </td>
      <td>{path.milleage}</td>
      <td
        className={!matchNames.first && matchNames.fuel ? 'inputError' : null}
        onDoubleClick={
          path.name !== null ? doubleClick('fuelBegin', path.fuelBegin) : null
        }
      >
        {path.name !== null ? (
          <span className="editableTd">{path.fuelBegin}</span>
        ) : null}
      </td>
      <td
        onDoubleClick={
          path.name !== null ? doubleClick('addFuel', path.addFuel) : null
        }
      >
        {path.name !== null ? (
          <span className="editableTd">{path.addFuel}</span>
        ) : (
          <span style={{ opacity: '0' }}>{'-'}</span>
        )}
      </td>
      <td>{path.deltaFuel}</td>
      <td className={matchNames.first && matchNames.fuel ? 'inputError' : null}>
        {path.fuelEnd}
      </td>
    </tr>
  );
};
export default RowTD;
