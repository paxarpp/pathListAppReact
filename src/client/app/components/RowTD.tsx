import React, { Fragment } from 'react';
import styled, { css, keyframes } from 'styled-components';
import Icon from './Icon';
import { IPath, ICar, IError } from './interfaces';

interface IProps {
  key: any;
  path?: IPath & ICar & IError;
  handler?: (p: IPath) => any;
  doubleClick?: (a: string, b: number) => any;
  deletePath: (p: IPath & ICar & IError) => any;
  selectPath: IPath;
};

const RowTD = ({ path, handler, doubleClick, deletePath, selectPath }: IProps) => (
  <WrapTr
    onClick={path.name !== null ? handler(path) : null}
    selected={selectPath && selectPath.dateBegin === path.dateBegin}
  >
    {path.dateBegin === null ? (
      <td />
    ) : (
      <td>{new Date(path.dateBegin).toLocaleDateString()}</td>
    )}
    <td>{path.fuel}</td>
    <td>
      {path.constFuelChange}
      <SpanExtension>
        {path.extension ? ' прицеп' : ''}
      </SpanExtension>
    </td>
    <td>
      {path.name !== null ? (
        <Fragment>
          <Span error={path.first && path.errorPath}>{path.pathBegin}</Span>
          <WrapIcon
            name="Create"
            color="green"
            size="16px"
            onClick={doubleClick('pathBegin', path.pathBegin)}
          />
        </Fragment>
      ) : null}
    </td>
    <td>
      {path.name !== null ? (
        <Fragment>
          <Span error={path.last && path.errorPath}>{path.pathEnd}</Span>
          <WrapIcon
            name="Create"
            color="green"
            size="16px"
            onClick={doubleClick('pathEnd', path.pathEnd)}
          />
        </Fragment>
      ) : null}
    </td>
    <td>{path.milleage}</td>
    <td>
      {path.name !== null ? (
        <Fragment>
          <Span error={path.first && path.errorFuel}>{path.fuelBegin}</Span>
          <WrapIcon
            name="Create"
            color="green"
            size="16px"
            onClick={doubleClick('fuelBegin', path.fuelBegin)}
          />
        </Fragment>
      ) : null}
    </td>
    <td>
      {path.name !== null ? (
        <span>
          {path.addFuel}
          <WrapIcon
            name="Create"
            color="green"
            size="16px"
            onClick={doubleClick('addFuel', path.addFuel)}
          />
        </span>
      ) : null}
    </td>
    <td>{path.deltaFuel}</td>
    <Td error={path.last && path.errorFuel}>{path.fuelEnd}</Td>
    <td>
      {path.name !== null ? (
        <Icon name="Delete" color="red" onClick={deletePath(path)} />
      ) : null}
    </td>
  </WrapTr>
);

const WrapIcon = styled(Icon)`
  position: absolute;
  top: 2px;
  right: 2px;
  transition: transform 0.2s linear;
  :hover {
    transform: scale(1.1);
  }
`;
const WrapTr = styled.tr<{selected: boolean}>`
  background-color: ${props => props.selected && 'antiquewhite'};
`;
const SpanExtension = styled.span`
  font-size: 0.7rem;
  color: #920000;
`;
const tremor = keyframes`
  0%{
transform: translateX(0);
  }
  25%{
    transform: translateX(3px);
  }
  50%{
    transform: translateX(0);
  }
  75%{
    transform: translateX(-3px);
  }
  100%{
    transform: translateX(0);
  }
`;
const isError = props =>
  props.error &&
  css<{error: boolean}>`
    color: red;
    font-weight: 600;
    display: inline-block;
    animation: ${tremor} 0.3s linear infinite;
  `;
const Span = styled.span`
  ${isError};
`;
const Td = styled.td<{error: boolean}>`
  color: ${({error}) => error && 'red'};
`;

export default RowTD;
