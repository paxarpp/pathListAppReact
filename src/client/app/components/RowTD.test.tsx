import React from 'react';
import renderer from 'react-test-renderer';
import RowTD from './RowTD';

test('RowTD view, not error not select', () => {
  const path = {
    ConsumptionFactoryFuel: 15.4,
    addFuel: 0,
    addFuelWinter: 0,
    constFuelChange: 15.4,
    dateBegin: '2018-05-03',
    deltaFuel: 11.24,
    fuel: 'AI',
    fuelBegin: 40.47,
    fuelEnd: 29.23,
    milleage: 73,
    name: 'уаз',
    pathBegin: 116239,
    pathEnd: 116312,
    first: '',
    errorFuel: '',
    errorPath: '',
  };
  const component = renderer.create(
    <RowTD
      handler={jest.fn()}
      doubleClick={jest.fn()}
      deletePath={jest.fn()}
      selectPath={{}}
      path={path}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('RowTD view, error not select', () => {
  const path = {
    ConsumptionFactoryFuel: 15.4,
    addFuel: 0,
    addFuelWinter: 0,
    constFuelChange: 15.4,
    dateBegin: '2018-05-03',
    deltaFuel: 11.24,
    fuel: 'AI',
    fuelBegin: 40.47,
    fuelEnd: 29.23,
    milleage: 73,
    name: 'уаз',
    pathBegin: 116239,
    pathEnd: 116312,
    errorFuel: true,
    errorPath: false,
    first: true,
  };
  const component = renderer.create(
    <RowTD
      handler={jest.fn()}
      doubleClick={jest.fn()}
      deletePath={jest.fn()}
      selectPath={{}}
      path={path}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('RowTD view, error select', () => {
  const path = {
    ConsumptionFactoryFuel: 15.4,
    addFuel: 0,
    addFuelWinter: 0,
    constFuelChange: 15.4,
    dateBegin: '2018-05-03',
    deltaFuel: 11.24,
    fuel: 'AI',
    fuelBegin: 40.47,
    fuelEnd: 29.23,
    milleage: 73,
    name: 'уаз',
    pathBegin: 116239,
    pathEnd: 116312,
    errorFuel: true,
    errorPath: false,
    first: true,
  };
  const selectPath = {
    ConsumptionFactoryFuel: 15.4,
    addFuel: 0,
    addFuelWinter: 0,
    constFuelChange: 15.4,
    dateBegin: '2018-05-03',
    deltaFuel: 11.24,
    fuel: 'AI',
    fuelBegin: 40.47,
    fuelEnd: 29.23,
    milleage: 73,
    name: 'уаз',
    pathBegin: 116239,
    pathEnd: 116312
  };
  const component = renderer.create(
    <RowTD
      handler={jest.fn()}
      doubleClick={jest.fn()}
      deletePath={jest.fn()}
      selectPath={selectPath}
      path={path}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
