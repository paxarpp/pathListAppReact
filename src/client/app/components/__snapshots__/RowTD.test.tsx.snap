// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowTD view, error not select 1`] = `
<tr
  className="sc-htpNat jpjkcF"
  selected={false}
>
  <td>
    2018-5-3
  </td>
  <td>
    AI
  </td>
  <td>
    15.4
    <span
      className="sc-bxivhb illsPj"
    >
      
    </span>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116239
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116312
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    73
  </td>
  <td>
    <span
      className="sc-ifAKCX hNONCD"
    >
      40.47
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span>
      0
      <svg
        className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
        fill="green"
        height="16px"
        viewBox="0 0 24 24"
        width="16px"
        x="0px"
        y="0px"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
  </td>
  <td>
    11.24
  </td>
  <td
    className="sc-EHOje eylSLQ"
  >
    29.23
  </td>
  <td>
    <svg
      className="sc-bdVaJa dgUyFF"
      fill="red"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      x="0px"
      y="0px"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
      />
    </svg>
  </td>
</tr>
`;

exports[`RowTD view, error select 1`] = `
<tr
  className="sc-htpNat fRsJUA"
  selected={true}
>
  <td>
    2018-5-3
  </td>
  <td>
    AI
  </td>
  <td>
    15.4
    <span
      className="sc-bxivhb illsPj"
    >
      
    </span>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116239
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116312
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    73
  </td>
  <td>
    <span
      className="sc-ifAKCX hNONCD"
    >
      40.47
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span>
      0
      <svg
        className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
        fill="green"
        height="16px"
        viewBox="0 0 24 24"
        width="16px"
        x="0px"
        y="0px"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
  </td>
  <td>
    11.24
  </td>
  <td
    className="sc-EHOje eylSLQ"
  >
    29.23
  </td>
  <td>
    <svg
      className="sc-bdVaJa dgUyFF"
      fill="red"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      x="0px"
      y="0px"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
      />
    </svg>
  </td>
</tr>
`;

exports[`RowTD view, not error not select 1`] = `
<tr
  className="sc-htpNat jpjkcF"
  selected={false}
>
  <td>
    2018-5-3
  </td>
  <td>
    AI
  </td>
  <td>
    15.4
    <span
      className="sc-bxivhb illsPj"
    >
      
    </span>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116239
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      116312
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    73
  </td>
  <td>
    <span
      className="sc-ifAKCX cpuQTc"
    >
      40.47
    </span>
    <svg
      className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
      fill="green"
      height="16px"
      viewBox="0 0 24 24"
      width="16px"
      x="0px"
      y="0px"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </td>
  <td>
    <span>
      0
      <svg
        className="sc-bwzfXH cxHokv sc-bdVaJa dgUyFF"
        fill="green"
        height="16px"
        viewBox="0 0 24 24"
        width="16px"
        x="0px"
        y="0px"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
    </span>
  </td>
  <td>
    11.24
  </td>
  <td
    className="sc-EHOje eylSLQ"
  >
    29.23
  </td>
  <td>
    <svg
      className="sc-bdVaJa dgUyFF"
      fill="red"
      height={24}
      viewBox="0 0 24 24"
      width={24}
      x="0px"
      y="0px"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
      />
    </svg>
  </td>
</tr>
`;
